apps/src/sketchlab/SketchlabView.tsx

```diff
@@ -41,34 +46,62 @@ const INITIAL_INFO_PANEL_WIDTH = 290;
 const MIN_WORKSPACE_WIDTH = 400;
 const INITIAL_WORKSPACE_WIDTH = 800;
 
+// TO DO: add Jira to handle bad extensions.
+const MIME_TO_EXT = {
+  'image/svg+xml': 'svg',
+  'image/png': 'png',
+  'image/jpeg': 'jpg',
+  'image/gif': 'gif',
+  'image/webp': 'webp',
+  'image/bmp': 'bmp',
+  'image/x-icon': 'ico',
+  'image/avif': 'avif',
+  'image/jfif': 'jfif',
+  'application/octet-stream': 'bin',
+};
+
 const DEBOUNCED_WORKSPACE_SERIALIZATION_MS = 500;
 
+async function uploadBase64ToUrl(
+  dataUrl: string,
+  uploadUrl: string,
+  mimeType: string
+): Promise<Response> {
+  const localResponse = await fetch(dataUrl);
+  const blob = await localResponse.blob();
+  const file = new File([blob], 'file', {
+    type: mimeType,
+  });
+
+  return await HttpClient.put(uploadUrl, file);
+}
+
+interface SerializedExcalidrawState {
+  elements: ExcalidrawElement[];
+  appState: AppState;
+  files: BinaryFiles;
+  externalFiles?: SketchlabExternalFiles;
```

Comment:
Structurally, I was a bit torn on whether to keep the "external file" info (ie, a S3 URL) in the same object with each file, or in a separate object.

I landed on having the external files URLs separate from the Excalidraw "files" themselves (rather than, say, having the S3 URL as a property on each file object), such that we could keep our special sauce entirely separate from what Excalidraw needs to function.

Reply:
I think having it be an extension of the main excalidraw state makes sense!